### PR TITLE
ferminet model prototype

### PR DESCRIPTION
**Pull request by shaipranesh2**
# Pull Request Template

## Description
Ferminet model prototype
Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/ferminet.py [hunk 1]
```diff
@@ -2,27 +2,268 @@
...
+    hessian = torch.autograd.functional.hessian(self.forward,
+                                                input,
+                                                retain_graph=True)
+
+    potential = self.calculate_potential()
+    return potential - 0.5 * (torch.sum(torch.diag(hessian, 0)) +
+                              torch.sum(torch.pow(jacobian, 2)))
+
+
+class FerminetModel(TorchModel):
```
Comment (tonydavis629) [inline deepchem/models/torch_models/ferminet.py:266]: As Bharath suggested, you can override whatver TorchModel method is preventing you from running ElectronSampler between iterations.

File: deepchem/models/torch_models/ferminet.py [hunk 2]
```diff
@@ -2,27 +2,268 @@
...
+      g_one_up = torch.mean(one_up, dim=0).flatten()
+      g_one_down = torch.mean(one_down, dim=0).flatten()
+      tmp_one_electron = torch.Tensor([])
+      tmp_two_electron = torch.Tensor([])
+      # spin-up electrons
+      for j in range(0, self.total_electron):
+        one_stream = one_electron[j].flatten()
+        two_stream = two_electron[j]
+        g_two_up = torch.mean(two_stream[:self.spin[0]], dim=0).flatten()
+        g_two_down = torch.mean(two_stream[self.spin[0]:], dim=0).flatten()
```
Comment (tonydavis629) [inline deepchem/models/torch_models/ferminet.py:168]: isn't self.spin[0] for up electrons? Is this correct for g_two_down?

Comment (shaipranesh2) [reply deepchem/models/torch_models/ferminet.py:168]: the two_stream tensor consists of both up-spin and down-spin, so we are going to divide it here for calculating g_two_up and g_two_down. So, using ":self.spin[0]" ensures the index 0 to self.spin[0] will be in up-spin and "self.spin[0]:" will ensure the remaining electrons which are down-spin is taken into account

**Comment by shaipranesh2** [comment]
@tonydavis629 @peastman Did some final round of changes in the network , sampler and kfac. Can you guys go through it and see whether its fine?

**Comment by shaipranesh2** [comment]
@tonydavis629 @peastman To be more specific, after the pretraining for a 400 iterations, my model outputs an energy of -4.1931 hartrees, whereas it should be around -1 hartrees. One reason I think this is happening is because the ferminet implementation has different envelopes than mentioned in the paper to calculate the orbitals ([ferminet envelope](https://github.com/deepmind/ferminet/blob/68a58dd8173dc2d390a8cb21ab480f591395638f/ferminet/envelopes.py)) , and has a very less loss from the start. 

In general it would be nice to see your feedback on this issue and review my code, if I have matched what the paper have said. 😅 It's a long PR, but i guess once this issue is rectified it's good to go.

PS. Small batch size does not seem to be an issue in deepmind's implementation, and it give more or less about the same energy value around -1.0 hartrees

**Comment by tonydavis629** [comment]
It's going to be very hard to make a valid review of this PR, it's just too big. I would definitely recommend separating the optimizer, sampler, and model into separate PRs. Then we can merge in the optimizer and sampler assuming there are no issues, and more easily drill down into why the model doesn't perform as expected.

**Comment by shaipranesh2** [comment]
> It's going to be very hard to make a valid review of this PR, it's just too big. I would definitely recommend separating the optimizer, sampler, and model into separate PRs. Then we can merge in the optimizer and sampler assuming there are no issues, and more easily drill down into why the model doesn't perform as expected.

Sure will do that, and also i think the sampler and optimizer works good, only made some minor corrections to it and passes its tests on my local machine.

**Comment by shaipranesh2** [comment]
@tonydavis629 @peastman @rbharath Made 2 more smaller PRs #3201 #3200 . This PR contains changes made to ferminet.py and test_ferminet.py, while the other edits are yapf changes.
